receipt_upload/receipt_upload/line_items/semantic_proposer.py

```
@@ -0,0 +1,139 @@
+    min_similarity: float = 0.5,
+) -> List[ReceiptWordLabel]:
+    """Propose PENDING PRODUCT_NAME labels for unlabeled in-band words.
+
+    Args:
+        words: ReceiptWord entities (need ``.bounding_box``/``.text``/ids).
+        existing_labels: labels already on the receipt (anchors + any proposals
+            from the geometry pass); only unlabeled words are candidates.
+        words_client: ChromaClient exposing ``.query(collection_name=..., ...)``.
+        word_embeddings: cached embeddings keyed by ``(line_id, word_id)``.
+    """
+    labeled = {(l.line_id, l.word_id) for l in existing_labels}
```

Ruff (0.15.17)

[error] Ambiguous variable name: `l`

(E741)